### PR TITLE
Prevents King's Psychic Summon from summoning xenos in other z levels

### PR DESCRIPTION
**Pull request by lKiseki**
## About The Pull Request
Adds a check in the Psychic Summon ability that prevents xenos in a different z level from being summoned.

## Why It's Good For The Game
Two reasons:
- King being able to summon other xenos in different Z levels lends itself to some serious cheese -- be it bypassing a FOB siege entirely, easy shipside sweeps, or cases where xenos are stuck shipside with no way of getting back groundside while marines are also groundside (bonus points if you nuke).
- From an administrative standpoint, the admin team has had to deal with various cases where Psychic Summon was used in this manner. The headmins have ultimately decided that it is best to deal with this by preventing it from happening in the first place via code.

## Changelog
:cl: Lewdcifer
balance: Psychic Summon can no longer summon xenos in a different Z level.
/:cl:

### CODE REVIEW

**Comment by blackdav123** [comment]
king summon does have some special use cases where multi z is beneficial like in cases where xenos get caught on the alamo so the king summons to bring them back.

leaving the functionality of pulling xenos back to groundside should stay and wouldnt impact the removal of pulling xenos to shipshide

**Review by Lumipharon — CHANGES_REQUESTED** [review]
I'm happy with this, although there will be some edge case issues with how you've done this, if xenos change z level mid summon (i.e. alamo/tad).

Make the actual forcemove be after the check, but the remove filter should happen regardless.
You should also purge xenos out of allxenos list in the first check, so you don't get funny edge cases oh no warning summon.

**Comment by Losenis** [comment]
> or cases where xenos are stuck shipside with no way of getting back groundside

I've seen more cases where summon _prevented_ this by having the xenos that got caught in the alamo midlaunch thinking it was hijack rather than normal cycle be brought down rather than cases where king moved them shipside, in all honesty.

**Comment by Oscari-pog** [comment]
Salt PR?

**Review by TiviPlus — APPROVED** [review]

**Comment by ivanmixo** [comment]
> I'm happy with this, although there will be some edge case issues with how you've done this, if xenos change z level mid summon (i.e. alamo/tad).
> 
> Make the actual forcemove be after the check, but the remove filter should happen regardless. You should also purge xenos out of allxenos list in the first check, so you don't get funny edge cases oh no warning summon.

What? Just register a signal to interrupt the ability cast if the z changes.

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 1]
```diff
@@ -485,16 +485,22 @@
 	xeno_message("King: \The [owner] has begun a psychic summon in <b>[get_area(owner)]</b>!", hivenumber = X.hivenumber)
 	var/list/allxenos = X.hive.get_all_xenos()
 	for(var/mob/living/carbon/xenomorph/sister AS in allxenos)
+		if(sister.z != owner.z)
+			continue
 		sister.add_filter("summonoutline", 2, outline_filter(1, COLOR_VIOLET))
 
 	if(!do_after(X, 10 SECONDS, FALSE, X, BUSY_ICON_HOSTILE))
```
Comment (ivanmixo) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:492]:
```suggestion
	if(!do_after(X, 10 SECONDS, FALSE, X, BUSY_ICON_HOSTILE))
```
Like I'd add a callback or something to this.

Comment (Lumipharon) [reply code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:492]: why would you want the teleport to cancel if one random xeno changes z-level?

**Comment by Oscari-pog** [comment]
Wouldn't the administrative issues not matter if the Queen allows the King to summon? The hive is still not a democracy.

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm (hunk 1 above)
Comment (ivanmixo) [reply code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:492]: If **you** change Z level

Comment (Lumipharon) [reply code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:492]: That's not what this is checking tho

**Comment by Autisem** [comment]
>be it bypassing a FOB siege entirely

Skill issue honestly. its funny when it happens

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 2]
```diff
@@ -485,16 +485,22 @@
...
 		add_cooldown(5 SECONDS)
 		for(var/mob/living/carbon/xenomorph/sister AS in allxenos)
+			if(sister.z != owner.z)
+				continue
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:496]:
```suggestion
```
Fail state, just removes the outline

File: code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm [hunk 3]
```diff
@@ -485,16 +485,22 @@
...
 			sister.remove_filter("summonoutline")
 		return fail_activate()
 
 	allxenos = X.hive.get_all_xenos() //refresh the list to account for any changes during the channel
 	for(var/mob/living/carbon/xenomorph/sister AS in allxenos)
+		if(sister.z != owner.z)
+			continue
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/king/abilities_king.dm:503]: outline should be always removed (xeno on z, moves away, will keep outline forever)

**Review by TiviPlus — CHANGES_REQUESTED** [review]
Otherwise lgtm

**Comment by chizzy376** [comment]
Please merge, this ability is extremely unfun to play against.

**Pull request closed without merging**